plugin/route53/route53.go

```diff
@@ -30,28 +30,44 @@ type Route53 struct {
 	upstream  *upstream.Upstream
 
 	zMu   sync.RWMutex
-	zones map[string]*zone
+	zones zones
 }
 
 type zone struct {
-	id string
-	z  *file.Zone
+	id  string
+	z   *file.Zone
+	dns string
+}
+
+type zones map[string][]*zone
+
+// AddHostedZone adds a new hosted zone into the zones map.
+// Returns true if the zone introduces a new domain.
+func (z zones) AddHostedZone(hz *zone) (newDomain bool) {
+	hz.z = file.NewZone(hz.dns, "")
+	if _, ok := z[hz.dns]; !ok {
+		z[hz.dns] = make([]*zone, 0)
+		newDomain = true
+	}
+	z[hz.dns] = append(z[hz.dns], hz)
+	return
 }
 
 // New returns new *Route53.
-func New(ctx context.Context, c route53iface.Route53API, keys map[string]string, up *upstream.Upstream) (*Route53, error) {
-	zones := make(map[string]*zone, len(keys))
+func New(ctx context.Context, c route53iface.Route53API, keys []*zone, up *upstream.Upstream) (*Route53, error) {
+	zones := make(zones, len(keys))
 	zoneNames := make([]string, 0, len(keys))
-	for dns, id := range keys {
+	for _, hostedZone := range keys {
 		_, err := c.ListHostedZonesByNameWithContext(ctx, &route53.ListHostedZonesByNameInput{
-			DNSName:      aws.String(dns),
-			HostedZoneId: aws.String(id),
+			DNSName:      aws.String(hostedZone.dns),
+			HostedZoneId: aws.String(hostedZone.id),
 		})
 		if err != nil {
 			return nil, err
 		}
-		zones[dns] = &zone{id: id, z: file.NewZone(dns, "")}
-		zoneNames = append(zoneNames, dns)
+		if zones.AddHostedZone(hostedZone) {
```

Comment:
Kind of prefer the old `keys` rather than passing half-initialized structures around. Probably need to invert it though to allow for multiple ids for the same zone name. Also could you document input params in the docstring please.

Reply:
just inverting was my initial thought, however it was ending up with losing the hosted zone ordering.

I changed the structure to a map[string][]string instead where the key is domain name and the value is a list of hosted zone ids. Let me know how you feel about it.